README.md
[![Build Status](https://api.travis-ci.org/kolypto/py-smsframework-vianett.png?branch=master)](https://travis-ci.org/kolypto/py-smsframework-vianett)

SMSframework Vianett Provider
================================

[Vianett](http://www.vianett.com/) Provider for [smsframework](https://pypi.python.org/pypi/smsframework/).

You need an account with "SMS Server" service set up.
You'll need the following configuration: username, password.






Installation
============

Install from pypi:

    $ pip install smsframework_vianett

To receive SMS messages, you need to ensure that
[Flask microframework](http://flask.pocoo.org) is also installed:


    $ pip install smsframework_vianett[receiver]






Initialization
==============

```python
from smsframework import Gateway
from smsframework_vianett import VianettProvider

gateway = Gateway()
gateway.add_provider('vianett', VianettProvider,
    user='kolypto',
    password='123',
    https=False,
    use_prefix=True
)
```

Config
------

Source: /smsframework_vianett/provider.py

* `user: str`: Account username
* `password: str`: Account password
* `https: bool`: Use HTTPS for outgoing messages? Default: `False`
* `use_prefix: bool`: Do you use prefixes for incoming messages?

    Stupidly, Vianett splits all incoming messages by space, and the first part goes to 'Prefix'.
    If you do not use prefixes, this can be very annoying!
    Set `False`: then, the whole message contents goes to 'body'.






Sending Parameters
==================

Provider-specific sending params: None






Additional Information
======================

OutgoingMessage.meta
--------------------
None.

IncomingMessage.meta
--------------------
* `prefix: str`: The first word in the message (keyword).
* `retrycount: int`: How many times the message has tried to be delivered.
* `operator: int`: The operator ID.
* `replypathid: int`: Only used for two-way dialogue, default 0.

MessageStatus.meta
------------------
... Tons of stupid, unpredictable fields






Receivers
=========

Source: /smsframework_vianett/receiver.py

Message Receiver: /im
---------------------
Go to Configuration > Connections, click 'Change'.
Put the message receiver URL into "HTTP url" field.

Message Receiver URL: `<provider-name>/im`

Status Receiver: /status
------------------------
Go to Configuration > Connections, click 'Change'.
Put the message receiver URL into "HTTP Status url" field.

Status Receiver URL: `<provider-name>/status`



CHANGES.md
Changelog
=========

v0.0.1, 2014.07.16
------------------

Initial release


test.py
#! /usr/bin/env python

from smsframework import Gateway
from smsframework_vianett import VianettProvider
from smsframework import OutgoingMessage

gateway = Gateway()
gateway.add_provider('vianett', VianettProvider,
    user='hakon@dignio.com',
    password='cn7xi',
    https=False
)

msg = OutgoingMessage('4797097418', 'Hello')
msg.provider_options.senderId = 'Dignio'
gateway.send(msg)


tests/vianett_test.py
# -*- coding: utf-8 -*-

import unittest
from datetime import datetime

from flask import Flask
from freezegun import freeze_time

from smsframework import Gateway, OutgoingMessage
from smsframework.providers import NullProvider
from smsframework_vianett import VianettProvider

from smsframework_vianett import error


class VianettProviderTest(unittest.TestCase):
    def setUp(self):
        # Gateway
        gw = self.gw = Gateway()
        gw.add_provider('null', NullProvider)  # provocation
        gw.add_provider('main', VianettProvider, user='kolypto', password='1234')

        # Flask
        app = self.app = Flask(__name__)

        # Register receivers
        gw.receiver_blueprints_register(app, prefix='/a/b/')

    def _mock_response(self, refno, errorcode, text):
        """ Monkey-patch VianettHttpApi so it returns a predefined response """
        def _api_request(method, **params):
            return '<?xml version="1.0"?><ack refno="{}" errorcode="{}">{}</ack>'.format(refno, errorcode, text)
        self.gw.get_provider('main').api._api_request = _api_request

    def test_blueprints(self):
        """ Test blueprints """
        self.assertEqual(
            self.gw.receiver_blueprints().keys(),
            ['main']
        )

    def test_api_request(self):
        """ Test raw requests """
        provider = self.gw.get_provider('main')

        # OK
        self._mock_response(1, '200', 'OK')
        self.assertEqual(provider.api_request('MT'), {'refno': '1', 'errorcode': '200', 'text': 'OK'})

        # Error reported
        self._mock_response(2, '400', 'Fail')
        self.assertRaises(error.VianettProviderError, provider.api_request, 'MT')

    def test_send(self):
        """ Test message send """
        gw = self.gw

        # OK
        self._mock_response(11111111, '200', 'OK')
        message = gw.send(OutgoingMessage('+123456', 'hey', provider='main'))
        self.assertEqual(message.msgid, '11111111')

        # Failure
        self._mock_response(22222222, '400', 'FAIL')
        self.assertRaises(error.VianettProviderError, gw.send, OutgoingMessage('+123456', 'hey', provider='main'))

    @freeze_time('2014-07-01 12:00:00')
    def test_receive_message(self):
        """ Test message receipt """

        # Message receiver
        messages = []
        def receiver(message):
            messages.append(message)
        self.gw.onReceive += receiver

        with self.app.test_client() as c:
            res = c.get('/a/b/main/im'
                        '?refno=19194091'
                        '&now=20140623122057'
                        '&requesttype=mo'
                        '&sourceaddr=47580008000626'
                        '&destinationaddr=4794041334'
                        '&replypathid=0'
                        '&prefix=TEST'
                        '&message=Hi,%20man'
                        '&retrycount=0'
                        '&operator=435'
                        '&referenceid='
                        '&username='
                        '&password=')
            self.assertEqual(res.status_code, 200)
            self.assertEqual(len(messages), 1)
            message = messages.pop()
            self.assertEqual(message.provider, 'main')
            self.assertEqual(message.msgid, '19194091')
            self.assertEqual(message.src, '47580008000626')
            self.assertEqual(message.dst, '4794041334')
            self.assertEqual(message.body, 'Hi, man')
            self.assertEqual(message.rtime.isoformat(), '2014-07-01T12:00:00')
            self.assertEqual(message.meta, {'operator': '435', 'prefix': 'TEST', 'replypathid': '0', 'retrycount': '0'})

    @freeze_time('2014-07-01 12:00:00')
    def test_receive_status(self):
        """ Test status receipt """

        # TODO: this test only contains data from the documentation. It's probably wrong! test with real data instead

        # Status receiver
        statuses = []

        def receiver(status):
            statuses.append(status)

        self.gw.onStatus += receiver

        with self.app.test_client() as c:
            # Status 1: notificationstatus, ACCEPTD
            res = c.get('/a/b/main/status'
                        '?password='
                        '&username='
                        '&refno=1234'
                        '&Status=ACCEPTD'
                        '&requesttype=notificationstatus'
                        '&StatusDescription=Absent+subscriber'
                        '&StatusCode=107'
                        '&now=06%2E10%2E2005+11%3A24%3A07&')
            self.assertEqual(res.status_code, 200)
            self.assertEqual(len(statuses), 1)
            st = statuses.pop()
            self.assertEqual(st.msgid, '1234')
            self.assertEqual(st.rtime.isoformat(), '2014-07-01T12:00:00')
            self.assertTrue('now' in st.meta)  # everything copied
            self.assertEqual(st.provider, 'main')
            self.assertEqual(st.accepted, True)
            self.assertEqual(st.delivered, False)
            self.assertEqual(st.expired, False)
            self.assertEqual(st.status_code, '107')
            self.assertEqual(st.status, 'ACCEPTD: Absent subscriber')
            self.assertEqual(st.error, False)

            # Status 2: notificationstatus, DELIVRD
            res = c.get('/a/b/main/status'
                        '?password='
                        '&username='
                        '&refno=1234'
                        '&Status=DELIVRD'
                        '&requesttype=notificationstatus'
                        '&StatusDescription='
                        '&StatusCode=0'
                        '&now=05%2E10%2E2005+00%3A30%3A06&')
            self.assertEqual(res.status_code, 200)
            self.assertEqual(len(statuses), 1)
            st = statuses.pop()
            self.assertEqual(st.msgid, '1234')
            self.assertEqual(st.rtime.isoformat(), '2014-07-01T12:00:00')
            self.assertTrue('now' in st.meta)  # everything copied
            self.assertEqual(st.provider, 'main')
            self.assertEqual(st.accepted, True)
            self.assertEqual(st.delivered, True)
            self.assertEqual(st.expired, False)
            self.assertEqual(st.status_code, '0')
            self.assertEqual(st.status, 'DELIVRD: ')
            self.assertEqual(st.error, False)

            # Status 3: Simple delivery report from operator:
            res = c.get('/a/b/main/status'
                        '?password='
                        '&username='
                        '&refno=1'
                        '&requesttype=mtstatus'
                        '&msgok=-1'
                        '&errorcode=0'
                        '&now=05%2E10%2E2005+03%3A04%3A57&')
            self.assertEqual(res.status_code, 200)
            self.assertEqual(len(statuses), 1)
            st = statuses.pop()
            self.assertEqual(st.msgid, '1')
            self.assertEqual(st.rtime.isoformat(), '2014-07-01T12:00:00')
            self.assertTrue('now' in st.meta)  # everything copied
            self.assertEqual(st.provider, 'main')
            self.assertEqual(st.accepted, True)
            self.assertEqual(st.delivered, True)
            self.assertEqual(st.expired, False)
            self.assertEqual(st.status_code, '0')
            self.assertEqual(st.status, '0 and -1')
            self.assertEqual(st.error, False)


            # Status 4: Advanced delivery report from operator
            res = c.get('/a/b/main/status'
                        '?password='
                        '&username='
                        '&refno=1234'
                        '&requesttype=mtstatus'
                        '&msgok=True'
                        '&ErrorDescription=OK'
                        '&ErrorCode=200'
                        '&Status='
                        '&SentDate=05%2E10%2E2005+00%3A41%3A27'
                        '&OperatorID=1'
                        '&CountryID=1'
                        '&CampaignID=1234'
                        '&Cut=90&CPAContentCost=0'
                        '&CPACost=0'
                        '&CPARevenue=1%2C35'
                        '&NetPrice=1%2C5'
                        '&ConsumerPrice=3'
                        '&PriceGroup=300'
                        '&Msg=Your message is accepted%2E'
                        '&Tel=1234'
                        '&FromAlpha=vianett'
                        '&sno=1963'
                        '&now=05%2E10%2E2005+00%3A41%3A43&')
            self.assertEqual(res.status_code, 200)
            self.assertEqual(len(statuses), 1)
            st = statuses.pop()
            self.assertEqual(st.msgid, '1234')
            self.assertEqual(st.rtime.isoformat(), '2014-07-01T12:00:00')
            self.assertTrue('now' in st.meta)  # everything copied
            self.assertEqual(st.provider, 'main')
            self.assertEqual(st.accepted, True)
            self.assertEqual(st.delivered, True)
            self.assertEqual(st.expired, False)
            self.assertEqual(st.status_code, '200')
            self.assertEqual(st.status, 'OK: : Your message is accepted.')
            self.assertEqual(st.error, False)


smsframework_vianett/api.py
# -*- coding: utf-8 -*-

import urllib
import urllib2
from xml.etree import ElementTree

from datetime import datetime

from . import const


class VianettApiError(RuntimeError):
    """ Error reported by Vianett """
    def __init__(self, code, message):
        self.code = code
        super(VianettApiError, self).__init__(message)


class VianettHttpApi(object):
    """ Vianett HTTP API client """

    def __init__(self, user, password, https=False):
        """ Create an authenticated client

            :param user: Authentication: username
            :param password: Authentication: password
            :type https: bool
            :param https: Use HTTPS protocol for requests?
        """
        self._auth = dict(
            username=user,
            password=password
        )
        self._https = https

        #: Provider API endpoint
        self._hostname = 'smsc.vianett.no'

    def _api_request(self, method, **params):
        """ Make an API request and return the result

            :rtype: str
        """
        url = {
            'MT': '{schema}://{host}/V3/CPA/MT/MT.ashx',  # Outgoing message
        }[method]

        # Prepare the request
        url = url.format(
            schema='https' if self._https else 'http',
            host=self._hostname,
        )
        data = {}
        data.update(self._auth)
        data.update(params)
        post = urllib.urlencode(data)

        # Request
        req = urllib2.Request(url, post)
        res = urllib2.urlopen(req)
        return res.read()

    def api_request(self, method, **params):
        """ Make a custom request to Vianett and get the response object.

            This also handles errors reported by the Vianett API

            :type method: str
            :param method: Method name to call
            :param params: Method parameters to send
            :rtype: dict
            :raises HTTPError: Http error code
            :raises URLError: Connection failed
            :raises AssertionError: Invalid response
            :raises VianettApiError: Vianett error
        """
        response = self._api_request(method, **params)

        # Parse
        try:
            root = ElementTree.fromstring(response)
        except ElementTree.ParseError as e:
            raise AssertionError('Failed to parse response: {}: {}'.format(e.message, response))
        assert root.tag == 'ack', 'Invalid response: {}'.format(response)

        # Error?
        if root.attrib['errorcode'] != '200':
            raise VianettApiError(root.attrib['errorcode'], root.text)

        # Okay
        ret = root.attrib
        ret['text'] = root.text
        return ret

    def sendmsg(self, to, text, **params):
        """ Send SMS message

            See :meth:`VianettHttpApi.api_request` for the list of raised exceptions.

            :param to: Destination number, digits only
            :param text: Message text: str or unicode.
            :param params: Message parameters. See Vianett docs.

            :param msgid: Message reference id. Default: (generated)
            :param SenderAddress: Sender address: SenderID, or a phone number. Default: None
            :param SenderAddressType: MSISN=1, short code=2, alphanumeric=5
            :param Priority: 0 -- low priority, >0 -- high priority

            :rtype: str
            :returns: Message id
        """
        # Params
        params.update(
            tel=to,
            msg=text
        )

        # msgid
        if 'msgid' not in params:
            params['msgid'] = datetime.now().strftime('%Y%m%d%H%M%S')

        # Sender
        if 'SenderAddress' in params and params['SenderAddress'].lstrip('+').isdigit() == False:
            params['SenderAddressType'] = const.SenderAddressType.ALPHANUMERIC

        # Send it, response
        res = self.api_request('MT', **params)
        return res['refno']


smsframework_vianett/provider.py
from smsframework import IProvider, exc
from . import error
from .api import VianettHttpApi, VianettApiError
from urllib2 import URLError, HTTPError


class VianettProvider(IProvider):
    """ Vianett provider """

    def __init__(self, gateway, name, user, password, https=False, use_prefix=True):
        """ Configure Vianett provider

            :param user: Account username
            :param password: Account password
            :param https: Use HTTPS for outgoing messages?
            :param use_prefix: Do you use prefixes for incoming messages?
                    Stupidly, Vianett splits all incoming messages by space, and the first part goes to 'Prefix'.
                    If you do not use prefixes, this can be very annoying!
                    Set `False`: then, the whole message contents goes to 'body'
        """
        self.api = VianettHttpApi(user, password, https)
        self.use_prefix = use_prefix
        super(VianettProvider, self).__init__(gateway, name)

    def send(self, message):
        """ Send a message

            :type message: smsframework.data.OutgoingMessage.OutgoingMessage
            :rtype: OutgoingMessage
            """
        # Parameters
        params = {}
        if message.src:
            params['SenderAddress'] = message.src
        if message.provider_options.senderId:
            params['SenderAddress'] = message.provider_options.senderId
        if message.provider_options.escalate:
            params['Priority'] = 1
        if message.provider_options.allow_reply:
            params['ReplyPathValue'] = 60*24  # Should be enough?
        params.update(message.provider_params)

        # Send
        try:
            message.msgid = self.api.sendmsg(message.dst, message.body, **params)
            return message
        except AssertionError as e:
            raise exc.RequestError(e.message)
        except HTTPError as e:
            raise exc.MessageSendError(e.message)
        except URLError as e:
            raise exc.ConnectionError(e.message)
        except VianettApiError as e:
            raise error.VianettProviderError(e.code, e.message)

    def make_receiver_blueprint(self):
        """ Create the receiver blueprint """
        from . import receiver
        return receiver.bp

    # region Public

    def api_request(self, method, **params):
        """ Raw request to Vianett API

            :rtype: str
            :raises RequestError: Request error
            :raises ConnectionError: Connection error
            :raises MessageSendError: HTTP error
            :raises VianettProviderError: Error with the request
        """
        try:
            return self.api.api_request(method, **params)
        except AssertionError as e:
            raise exc.RequestError(e.message)
        except HTTPError as e:
            raise exc.MessageSendError(e.message)
        except URLError as e:
            raise exc.ConnectionError(e.message)
        except VianettApiError as e:
            raise error.VianettProviderError(e.code, e.message)

    #endregion


smsframework_vianett/receiver.py
from datetime import datetime, timedelta

from flask import Blueprint
from flask.globals import request, g

from smsframework.data import IncomingMessage
from smsframework.data import MessageAccepted, MessageDelivered, MessageExpired, MessageError

bp = Blueprint('smsframework-vianett', __name__, url_prefix='/')


@bp.route('/im')
def im():
    """ Incoming message handler

        * sourceaddr: The originator number.
        * destinationaddr: The destination number.
        * prefix: The first word in the message (keyword).
        * message: The last part of the message.
        * refno: ViaNett's message ID. Reference number.
        * retrycount: How many times the message has tried to be delivered.
        * operator: The operator ID. 1 is Telenor Mobile (Norway), 2 is NetCom (Norway), etc.
        * requesttype: This field will have the value "mo" on incoming messages.
        * username: Optional username.
        * password: Optional password.
        * replypathid: Only used for two-way dialogue, default 0.
    """
    req = request.args.to_dict()

    # Check
    for n in ('sourceaddr', 'message', 'refno', 'destinationaddr', 'prefix', 'retrycount', 'operator', 'replypathid'):
        assert n in req, 'Vianett message with missing "{}" field: {}'.format(n, req)

    # Prefixes
    if not g.provider.use_prefix:
        req['message'] = ' '.join(filter(lambda x: x, (req['prefix'], req['message'])))
        req['prefix'] = ''

    # IncomingMessage
    message = IncomingMessage(
        src=req['sourceaddr'],
        body=req['message'],
        msgid=req['refno'],
        dst=req['destinationaddr'],
        rtime=datetime.utcnow(),
        meta = {
            'prefix': req['prefix'],
            'retrycount': req['retrycount'],
            'operator': req['operator'],
            'replypathid': req['replypathid']
        }
    )

    # Process it
    " :type: smsframework.IProvider.IProvider "
    g.provider._receive_message(message)  # any exceptions will respond with 500, and Vianett will happily retry later

    # Ack
    return '<ack refno="{msgid}" errorcode="0" />'.format(msgid=message.msgid)


@bp.route('/status')
def status():
    """ Incoming status report

        * refno: msgid
        * now: Datetime in CET: "06.10.2005 11:24:07" (not always provided)
        * requesttype: type of status report:

            'mtstatus': Delivered to the terminal
                Simple delivery report from operator:
                    * errorcode: '0' if delivered fine
                    * msgok: Whether the message was successfully submitted. (not always provided)
                        'True' -- yes, '-1' -- pending

                Advanced delivery report from operator:
                    * SentDate: Datetime in CET when the message was sent: "06.10.2005 11:24:07"
                    * ErrorCode: error code. '200' is ok
                    * ErrorDescription: error description string
                    * Status: (mysterious status text)
                    * Msg: (even more mysterious status text)
                    * CPACost, CPARevenue, NetPrice, ConsumerPrice: message price
                    * ... more stupid fields

            'notificationstatus': Sent to phone
                * Status:
                    'ACCEPTD' or 'BUFFERD' -- queued
                    'DELIVRD' -- delivered to the terminal
                * StatusDescription: Description of the 'Status' field (not always provided)
                * StatusCode: Code representing the status of the message
    """
    req = request.args.to_dict()

    def _check_fields(*fields):
        for n in fields:
            assert n in req, 'Vianett status with missing "{}" field: {}'.format(n, req)

    _check_fields('requesttype')
    if req['requesttype'] == 'notificationstatus':
        _check_fields('Status', 'StatusDescription', 'StatusCode')

        # Create status
        status = {
            'DELIVRD': MessageDelivered,
            'ACCEPTD': MessageAccepted,
            'BUFFERD': MessageAccepted
        }[req['Status']](req['refno'], meta=req)
        status.status_code = req['StatusCode']
        status.status = '{0[Status]}: {0[StatusDescription]}'.format(req)
    elif req['requesttype'] == 'mtstatus':
        if 'ErrorCode' in req:
            _check_fields('ErrorCode', 'ErrorDescription', 'Status', 'Msg')

            # Create status
            status = {
                True: MessageDelivered,
                False: MessageError
            }[req['ErrorCode'] == '200'](req['refno'], meta=req)
            status.status_code = req['ErrorCode']
            status.status = '{0[ErrorDescription]}: {0[Status]}: {0[Msg]}'.format(req)
        else:
            _check_fields('errorcode')

            # Simple
            status = {
                True: MessageDelivered,
                False: MessageError
            }[req['errorcode'] == '0'](req['refno'], meta=req)
            status.status_code = req['errorcode']
            status.status = '{} and {}'.format(req['errorcode'], req['msgok'] if 'msgok' in req else '?')
    else:
        raise AssertionError('Vianett status with an unsupported `requesttype`: {}'.format(req))

    # Process it
    g.provider._receive_status(status)  # exception respond with http 500

    # Ack
    return '<?xml version="1.0"?><ack refno="1234" errorcode="0" />'
